packages/jupyter-ai/jupyter_ai/document_loaders/directory.py

```diff
@@ -1,54 +1,73 @@
-import logging
-from typing import Dict, List, Optional, Union
+from typing import List, Optional
+from pathlib import Path
+import hashlib
+import itertools
+
+import ray
 
 from langchain.document_loaders.base import BaseLoader
-from langchain.document_loaders.directory import FILE_LOADER_TYPE, _is_visible
-from langchain.document_loaders.unstructured import UnstructuredFileLoader
 from langchain.schema import Document
-from wcmatch.pathlib import Path
-
-
-logger = logging.getLogger(__name__)
+from langchain.document_loaders.directory import _is_visible
+from langchain.text_splitter import (
+    RecursiveCharacterTextSplitter, TextSplitter,
+)
 
+@ray.remote
+def path_to_doc(path):
+    with open(str(path), 'r') as f:
+        text = f.read()
+        m = hashlib.sha256()
+        m.update(text.encode('utf-8'))
+        metadata = {'path': str(path), 'sha256': m.digest(), 'extension': path.suffix}
+        return Document(page_content=text, metadata=metadata)
 
-class DirectoryLoader(BaseLoader):
-    """Loading logic for loading documents from a directory."""
+class ExcludePattern(Exception):
+    pass
+    
+def iter_paths(path, extensions, exclude):
+    for p in Path(path).rglob('*'):
+        if p.is_dir():
+            continue
+        if not _is_visible(p.relative_to(path)):
+            continue
+        try:
+            for pattern in exclude:
+                if pattern in str(p):
+                    raise ExcludePattern()
+        except ExcludePattern:
+            continue
+        if p.suffix in extensions:
+            yield p
 
+class RayRecursiveDirectoryLoader(BaseLoader):
+    
```

Comment:
High level comment: does it make sense to determine whether we index a file based purely on its file extension? This strategy doesn't appear to be sustainable, even in the short-term, since we would need to declare every single file extension that could conceivably be under a user's project root and be indexed by FAISS. I can see this being an immediate pain point for users.

One alternative is to simply allow indexing of any file encoded as text while ignoring binary files. There appear to be a [few simple Python solutions](https://stackoverflow.com/questions/898669/how-can-i-detect-if-a-file-is-binary-non-text-in-python) we can leverage here. What are your thoughts on this?

Reply:
I have done quite a few tests and the details of parsing and splitting make a significant different to the quality of the output. For example, when we are splitting Python or markdown we want to not cut functions/classes/paragraphs in half during splitting. And for notebooks, we first convert to markdown and then split the markdown. We can improve this over time, but I do think we need as high quality of parsing/splitting as possible.